### PR TITLE
BUG: fft.fht: set `u.imag[-1] = 0` only when `n` is even

### LINKED ISSUE #21661: BUG: fft.fht: should set `u.imag[-1] = 0` only when `n` is even
### Describe your issue.

In the source of scipy.fft.fht, in the function fhtcoeff(...) in  _fftlog_backend.py, one has to ensure that the array u has a u[-1] which is purely real when the data is of even size n. That requirement stems from the fact that one wants the fht transform of real data to end up real. The argument is spelled out in the original paper. 

However, the code sets u[-1].imag=0 even for odd n, which I'm pretty sure is wrong. So, I believe that on lines 94-95:

    # fix last coefficient to be real
    u.imag[-1] = 0
    
there should be a check to ensure that u.imag[-1]=0 only for even n. 

### Reproducing Code Example

```python
The issue has to do with mathematical correctness of the implemented algorithm. The code as it is doesn't crash.
```

### Error message

```shell
none
```

### SciPy/NumPy/Python version and system information

```shell
1.14.1
```

### PR DESCRIPTION
**Pull request by stassev**
Closes #21661.

### CODE REVIEW

**Review by lucascolley — APPROVED** [review]
LGTM! @ntessore mind confirming?

**Comment by ntessore** [comment]
Would it make sense to add a test for this?

File: scipy/fft/_fftlog_backend.py [hunk 1]
```diff
@@ -92,7 +92,8 @@ def fhtcoeff(n, dln, mu, offset=0.0, bias=0.0, inverse=False):
     np.exp(u, out=u)
 
     # fix last coefficient to be real
-    u.imag[-1] = 0
+    if (n%2==0):
```
Comment (yuanx749) [inline scipy/fft/_fftlog_backend.py:95]:
```suggestion
    if n % 2 == 0:
```
Small nitpick.

**Review by tylerjereddy — COMMENTED** [review]
Here's a sample test that fails before/passes after the patch; I bet there is something simpler, though this is at least similar to the author's original complaint.

<details>

```diff
--- a/scipy/fft/tests/test_fftlog.py
+++ b/scipy/fft/tests/test_fftlog.py
@@ -6,7 +6,7 @@ from scipy.fft._fftlog import fht, ifht, fhtoffset
 from scipy.special import poch
 
 from scipy.conftest import array_api_compatible
-from scipy._lib._array_api import xp_assert_close
+from scipy._lib._array_api import xp_assert_close, xp_assert_less
 
 pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends"),]
 skip_xp_backends = pytest.mark.skip_xp_backends
@@ -177,3 +177,19 @@ def test_array_like(xp, op):
          [[1.0, 1.0], [1.0, 1.0]],
          [[1.0, 1.0], [1.0, 1.0]]]
     xp_assert_close(op(x, 1.0, 2.0), op(xp.asarray(x), 1.0, 2.0))
+
+def test_gh_21661(xp):
+    mu = 0.0
+    r = np.logspace(-7, 1, 129)
+    dln = np.log(r[1] / r[0])
+    offset = fhtoffset(dln, initial=-6 * np.log(10), mu=mu)
+    k = np.exp(offset) / r[::-1]
+
+    def f(x, mu):
+        return x**(mu + 1)*np.exp(-x**2/2)
+
+    a_r = f(r, mu)
+    fht_val = fht(a_r, dln, mu=mu, offset=offset)
+    a_k = f(k, mu)
+    rel_err = abs((fht_val - a_k) / a_k).max()
+    xp_assert_less(rel_err, 7.28e+16)

```

</details>

**Comment by tylerjereddy** [comment]
One more note on that test, if you try to switch the `np`->`xp` for array API regression testing as well, you'll need a bit more creativity or to add some skips since i.e., `AttributeError: module 'array_api_strict' has no attribute 'logspace'.`.

**Comment by lucascolley** [comment]
> no attribute 'logspace'

Not too much creativity required - `xp.asarray(np.logspace(...))` is fine.

**Comment by j-bowhay** [comment]
@lucascolley I was trying to add the patch but ran into difficulty with `fhtoffset`, does this support the array API? All the tests using it just seem to use numpy arrays

**Comment by lucascolley** [comment]
Right, `fhtoffset` is scalar input, scalar output, so the test can be `np_only` (and provide scalars rather than arrays).

It does use NumPy functions internally, but I'm not sure of the value of making it use a standard namespace instead.

**Comment by j-bowhay** [comment]
Thanks! Made some progress but still not quite working for torch:
```
diff --git a/scipy/fft/tests/test_fftlog.py b/scipy/fft/tests/test_fftlog.py
index 61a8ff892b..2737e3a4c8 100644
--- a/scipy/fft/tests/test_fftlog.py
+++ b/scipy/fft/tests/test_fftlog.py
@@ -6,7 +6,7 @@ from scipy.fft._fftlog import fht, ifht, fhtoffset
 from scipy.special import poch
 
 from scipy.conftest import array_api_compatible
-from scipy._lib._array_api import xp_assert_close
+from scipy._lib._array_api import xp_assert_close, xp_assert_less, array_namespace
 
 pytestmark = [array_api_compatible, pytest.mark.usefixtures("skip_xp_backends"),]
 skip_xp_backends = pytest.mark.skip_xp_backends
@@ -177,3 +177,21 @@ def test_array_like(xp, op):
          [[1.0, 1.0], [1.0, 1.0]],
          [[1.0, 1.0], [1.0, 1.0]]]
     xp_assert_close(op(x, 1.0, 2.0), op(xp.asarray(x), 1.0, 2.0))
+
+def test_gh_21661(xp):
+    xp_test = array_namespace(xp.asarray(1.0))
+    mu = 0.0
+    r = np.logspace(-7, 1, 129)
+    dln = np.log(r[1] / r[0])
+    offset = xp.asarray(fhtoffset(dln, initial=-6 * np.log(10), mu=mu))
+    r = xp.asarray(r)
+    k = xp.exp(offset) / xp_test.flip(r, axis=-1)
+
+    def f(x, mu):
+        return x**(mu + 1)*xp.exp(-x**2/2)
+
+    a_r = f(r, mu)
+    fht_val = fht(a_r, xp.asarray(dln), mu=xp.asarray(mu), offset=offset)
+    a_k = f(k, mu)
+    rel_err = xp.max(xp.abs((fht_val - a_k) / a_k))
+    xp_assert_less(rel_err, xp.asarray(7.28e+16)[()])
```

**Comment by lucascolley** [comment]
in the `fht` call, `dln`, `mu`, and `offset` should be scalars (if the current implementation is correct).

**Comment by j-bowhay** [comment]
> in the `fht` call, `dln`, `mu`, and `offset` should be scalars (if the current implementation is correct).

Ah yes thanks, it would help if I read the docs!

**Comment by j-bowhay** [comment]
> Right, `fhtoffset` is scalar input, scalar output, so the test can be `np_only` (and provide scalars rather than arrays).
> 
> It does use NumPy functions internally, but I'm not sure of the value of making it use a standard namespace instead.

Another reason we need #21255!!

**Comment by lucascolley** [comment]
> > Right, fhtoffset is scalar input, scalar output, so the test can be np_only (and provide scalars rather than arrays).
> > 
> > It does use NumPy functions internally, but I'm not sure of the value of making it use a standard namespace instead.
>
> Another reason we need https://github.com/scipy/scipy/issues/21255!!

Hmm, I don't think that documentation would have helped here - the array namespace used internally is an implementation detail, everything works fine if you provide Python scalars as the docs say.

**Review by lucascolley — APPROVED** [review]
LGTM!

File: scipy/fft/tests/test_fftlog.py [hunk 2]
```diff
@@ -177,3 +179,22 @@ def test_array_like(xp, op):
...
+
+def test_gh_21661(xp):
+    one = xp.asarray(1.0)
+    xp_test = array_namespace(one)
+    mu = 0.0
+    r = np.logspace(-7, 1, 129)
+    dln = np.log(r[1] / r[0])
+    offset = fhtoffset(dln, initial=-6 * np.log(10), mu=mu)
+    r = xp.asarray(r, dtype=one.dtype)
+    k = math.exp(offset) / xp_test.flip(r, axis=-1)
```
Comment (ntessore) [inline scipy/fft/tests/test_fftlog.py:191]: Given that `np.log()` is used three lines above, why `math.exp()` here?

**Comment by ntessore** [comment]
Thanks for adding the test, this should really have been in the original implementation.

Is there a chance you can parametrise the test for even/odd input? It should work well with both n=129 and n=128. Such a test would detect if the fix in this PR broke the behaviour for even lengths!

**Comment by stassev** [comment]
Regarding:

    rel_err = xp.max(xp.abs((fht_val - a_k) / a_k))

FHT returns garbage near the edges because of aliasing and ringing. That's just expected behavior. So, I think a fairer test would be to take the middle section of the relative error array, and take the xp.max only of that portion of the array.

**Comment by stassev** [comment]
> Thanks for adding the test, this should really have been in the original implementation.
> 
> Is there a chance you can parametrise the test for even/odd input? It should work well with both n=129 and n=128. Such a test would detect if the fix in this PR broke the behaviour for even lengths!

Also, maybe an add a test with and without bias (of say bias =-0.5), as that tests a separate part of the code that can break?

File: scipy/fft/tests/test_fftlog.py [hunk 3]
```diff
@@ -177,3 +179,22 @@ def test_array_like(xp, op):
          [[1.0, 1.0], [1.0, 1.0]],
          [[1.0, 1.0], [1.0, 1.0]]]
     xp_assert_close(op(x, 1.0, 2.0), op(xp.asarray(x), 1.0, 2.0))
+
+def test_gh_21661(xp):
+    one = xp.asarray(1.0)
+    xp_test = array_namespace(one)
+    mu = 0.0
+    r = np.logspace(-7, 1, 129)
+    dln = np.log(r[1] / r[0])
```
Comment (j-bowhay) [inline scipy/fft/tests/test_fftlog.py:188]:
```suggestion
    dln = math.log(r[1] / r[0])
```

File: scipy/fft/tests/test_fftlog.py [hunk 4]
```diff
@@ -177,3 +179,22 @@ def test_array_like(xp, op):
          [[1.0, 1.0], [1.0, 1.0]],
          [[1.0, 1.0], [1.0, 1.0]]]
     xp_assert_close(op(x, 1.0, 2.0), op(xp.asarray(x), 1.0, 2.0))
+
```
Comment (j-bowhay) [inline scipy/fft/tests/test_fftlog.py:182]:
```suggestion

@pytest.mark.parameterize('n', (128, 129))
```

File: scipy/fft/tests/test_fftlog.py [hunk 5]
```diff
@@ -177,3 +179,22 @@ def test_array_like(xp, op):
...
+def test_gh_21661(xp):
+    one = xp.asarray(1.0)
+    xp_test = array_namespace(one)
+    mu = 0.0
+    r = np.logspace(-7, 1, 129)
```
Comment (j-bowhay) [inline scipy/fft/tests/test_fftlog.py:187]:
```suggestion
    r = np.logspace(-7, 1, n)
```

File: scipy/fft/tests/test_fftlog.py [hunk 6]
```diff
@@ -177,3 +179,22 @@ def test_array_like(xp, op):
          [[1.0, 1.0], [1.0, 1.0]],
          [[1.0, 1.0], [1.0, 1.0]]]
     xp_assert_close(op(x, 1.0, 2.0), op(xp.asarray(x), 1.0, 2.0))
+
+def test_gh_21661(xp):
```
Comment (j-bowhay) [inline scipy/fft/tests/test_fftlog.py:183]:
```suggestion
def test_gh_21661(xp, n):
```

File: scipy/fft/tests/test_fftlog.py [hunk 7]
```diff
@@ -177,3 +179,23 @@ def test_array_like(xp, op):
          [[1.0, 1.0], [1.0, 1.0]],
          [[1.0, 1.0], [1.0, 1.0]]]
     xp_assert_close(op(x, 1.0, 2.0), op(xp.asarray(x), 1.0, 2.0))
+
+@pytest.mark.parameterize('n', [128, 129])
```
Comment (j-bowhay) [inline scipy/fft/tests/test_fftlog.py:183]:
```suggestion
@pytest.mark.parametrize('n', [128, 129])
```

**Review by j-bowhay — APPROVED** [review]
Ok let's get this in so it makes the next release, follow up pr's to improve the test further are very welcome.

Thanks for your first contribution to SciPy @stassev

**Pull request merged**